#### Aquí encontrarás todo lo que necesitas saber: https://www.kaggle.com/t/ab8726f0cfc84544abbae69a6be88071

In [22]:
import numpy as np
import pandas as pd


import urllib.request

import seaborn as sns
from sklearn import preprocessing
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn import svm
from sklearn import model_selection
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.pylab as pylab


In [23]:
### Para que funcione necesitas bajarte los archivos de datos de Kaggle 
df = pd.read_csv("diamonds_train.csv", index_col=0)

In [24]:
df.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 40345 entries, 0 to 40344
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   carat    40345 non-null  float64
 1   cut      40345 non-null  object 
 2   color    40345 non-null  object 
 3   clarity  40345 non-null  object 
 4   depth    40345 non-null  float64
 5   table    40345 non-null  float64
 6   x        40345 non-null  float64
 7   y        40345 non-null  float64
 8   z        40345 non-null  float64
 9   price    40345 non-null  int64  
dtypes: float64(6), int64(1), object(3)
memory usage: 3.4+ MB


In [25]:
df.shape

(40345, 10)

In [26]:
"""dispersiones = sns.pairplot(df)"""
#mirar imagenes de dispersiones, no puedo runnearlo aqui que me peta todo 


'dispersiones = sns.pairplot(df)'

In [27]:
#he entendido que si son las medidas del diamante, no puede haber un diamante que tenga 2 medidas, es decir que seria como en 2d 
df = df.drop(df[df["x"]==0].index)
df = df.drop(df[df["y"]==0].index)
df = df.drop(df[df["z"]==0].index)

In [28]:
df.shape

(40328, 10)

In [29]:
df = df[(df["table"]<80)&(df["table"]>40)]
df = df[(df["x"]<30)]
df = df[(df["y"]<30)]

df = df[(df["z"]<30)&(df["z"]>2)]

In [30]:
df.shape


(40322, 10)

In [31]:
#hay tres columnas con valores categoricos cut color y clarity 
df.dtypes

carat      float64
cut         object
color       object
clarity     object
depth      float64
table      float64
x          float64
y          float64
z          float64
price        int64
dtype: object

In [32]:
etiquetas = preprocessing.LabelEncoder()
categoricas = (df.dtypes =="object")
columnas_categoricas = list(categoricas[categoricas].index)

for col in columnas_categoricas:
    df[col] = etiquetas.fit_transform(df[col])

In [33]:
df.head()

carat  cut  color  clarity  depth  table     x     y     z  price
0   1.21    2      4        5   63.0   57.0  6.73  6.70  4.23   6134
1   0.28    4      0        7   64.0   56.0  4.14  4.17  2.66    532
2   0.42    3      2        4   61.2   58.0  4.86  4.82  2.96   1103
3   0.26    2      4        1   61.1   57.0  4.16  4.12  2.53    600
4   1.10    1      3        2   63.4   57.0  6.52  6.55  4.14   4997

In [34]:
"""correlaciones= df.corr()
ax = plt.subplots(figsize=(10,10))
sns.heatmap(correlaciones,annot=True)"""
#mirar imagen de correlaciones, me peta todo si lo hago aqui

'correlaciones= df.corr()\nax = plt.subplots(figsize=(10,10))\nsns.heatmap(correlaciones,annot=True)'

In [35]:
#viendo las correlaciones, solo me voy a quedarcon "carat","x","y" y "z" ya que son las que tienene unas correelaciones más fuertes. 

In [36]:
df.drop(["cut","color","clarity","depth","table"], axis=1, inplace=True)

In [37]:
df.info() #no hay valores nan

<class 'pandas.core.frame.DataFrame'>
Int64Index: 40322 entries, 0 to 40344
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   carat   40322 non-null  float64
 1   x       40322 non-null  float64
 2   y       40322 non-null  float64
 3   z       40322 non-null  float64
 4   price   40322 non-null  int64  
dtypes: float64(4), int64(1)
memory usage: 1.8 MB


In [38]:
#he querido estandarizar, pero me salia en array y a pablo las arrays no le gustan 

In [39]:
X= np.array(df.drop(["price"],axis =1)) #separo en X-y 
y= np.array(df["price"])

In [40]:
X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.2, random_state=7) #habemos partición 

In [41]:
polinominal_model = PolynomialFeatures(4) 
X_poly = polinominal_model.fit_transform(X, y)
X_train_poly, X_test_poly, y_train, y_test = train_test_split(X_poly, y, test_size = 0.2, random_state = 7)
lin_reg_model = LinearRegression()
lin_reg_model.fit(X_train_poly, y_train)
print ('score train :', lin_reg_model.score(X_train_poly, y_train))
print ('score test: ', lin_reg_model.score(X_test_poly, y_test))
print ('score total:', lin_reg_model.score(X_poly, y))

print ('coeficiente:', lin_reg_model.coef_)


score train : 0.8825499967775162
score test:  0.8781073099021789
score total: 0.8816629512858114
coeficiente: [-8.21063551e-04 -1.51811642e+06 -2.18775562e+05 -3.29675330e+05
  1.53097595e+06 -1.28716750e+06 -6.12838277e+04 -5.76135692e+05
  2.62289247e+06  3.37493117e+05 -5.18176759e+05  6.22394114e+04
  4.36072509e+05 -2.30240028e+05 -7.47706573e+05 -3.16633279e+05
 -1.69505152e+05 -5.86534456e+04  1.16670102e+06  4.90853199e+05
 -9.26902815e+05  1.42341903e+05  6.50069295e+05 -3.23884567e+05
 -7.87899717e+05 -4.99940060e+04  2.18313620e+05 -4.16732553e+05
 -3.96594731e+05  1.07179311e+06 -2.57011636e+05  2.04115288e+05
 -7.30453767e+05  4.26538360e+05  1.03804874e+05 -1.94753907e+04
 -8.15416184e+04  4.40239917e+04  1.46174967e+05  1.85255642e+05
 -3.58957236e+05  1.16194594e+05  2.16998305e+05 -1.52981036e+05
 -1.68512430e+05 -4.45841281e+04  1.58146093e+05 -2.63444648e+05
 -2.22935213e+05  6.28450475e+05 -1.45075921e+05  9.87523368e+04
 -3.94179381e+05  2.17269226e+05  5.69885871e

**Importante:**

   - Si quitas columnas o creas columnas nuevas a partir de otras, o cualquier modificación column-wise tendrás que aplicarlo al dataset de `diamonds_test.csv` de cara a hacer la predicción.
   - Si por lo contrario, decides por ejemplo, quitar los outliers o hacer un `dropna()`, o cualquier modificación row-wise eso NO PODRÁS (ni debes) aplicarlo al dataset de `diamonds_test.csv` de cara a hacer la predicción. ¿Por qué? Porque si el conjunto de test tiene 50 observaciones (filas) la predicción se espera que tenga 50 filas.

#### ¿Y si lo que hago es estandarizar los datos o hacer un encoding, también se lo tengo que hacer al test antes de predecir?

In [42]:
# ¿qué opináis?
#Sí, habría que hacerlo

-----------------------------------------------------------------------------------------------------------------

### Pasos que si o si debéis realizar para poder participar en la competición:

### 1. Definir X e y

In [43]:
# coger las columnas "x", "y", "z" tiene solo valor ilustrativo, NO LO COPIÉIS
X = np.array(df[["x","y","z"]])
y = np.array(df["price"])
X.shape

(40322, 3)

In [44]:
y.shape

(40322,)

#### ¿Y si después del análisis exploratorio de mis datos llego a la conclusión de que puedo predecir el precio solo con las columnas `x`, `y` y `z`, también aplica al test?

In [45]:
# ¿qué opináis?
# sí, el conjunto para predecir `diamonds_test.csv` también tendrá que tener 3 columnas

### 2. Dividir X_train, X_test, y_train, y_test

In [46]:
# el test_size 0.20 y el random_state puede cambiar
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 42)

In [47]:
X_train

array([[6.4 , 6.46, 3.85],
       [4.3 , 4.32, 2.66],
       [5.64, 5.76, 3.37],
       ...,
       [4.23, 4.29, 2.63],
       [6.5 , 6.53, 3.99],
       [7.4 , 7.44, 4.57]])

### 3. Asignar el modelo (vacío) a una variable
Aquí meteríais los parámetros. 

**Consejo**: Usa GridSearch y vuélvete loca o loco probando modelos e hiperparámetros.

In [48]:
model = LinearRegression()

Siempre hay tiempo para una historia:
https://catboost.ai/docs/concepts/python-reference_catboostregressor.html

### 4. Entrenar el modelo con X_train, y_train

In [49]:
model.fit(X_train, y_train)

LinearRegression()

### 5. Predecir con el modelo ya entrenado con X_test

In [50]:
predictions = model.predict(X_test)
print(predictions)

[7203.97822702 3748.71629064 6313.20580257 ...  790.11786339 8940.12415447
 2509.10325743]


### 6. Sacar métricas, valorar el modelo 

Recuerdo que en la competición se va a evaluar con la métrica de RMSE.

In [51]:
np.sqrt(mean_squared_error(y_test, predictions))

1869.6001808370793

-----------------------------------------------------------------

## Una vez listo el modelo, toca predecir con el dataset de predicción 

Definición de **modelo que está listo**. 

_Tras hacer suficientes pruebas, analizar los datos, hacer feature engineering, probar diferentes modelos con diferentes parámetros, es con este con el que observo mejores métricas y menos overfitting. ¡Cuidado con el overfitting aquí! Si vuestro modelo aprende muy bien de estos datos pero hay overfitting cuando le pasemos los datos desconocidos de `diamonds_test.csv` nos arriesgamos a que digamos, no salga lo esperado._

### 1. Entrena dicho modelo con TODOS tus datos de train, esto es con `diamonds_train.csv` al completo.

**CON LAS TRANSFORMACIONES QUE LE HAYAS REALIZADO A `X` INCLUÍDAS.**

Véase:
- Estandarización/Normalización
- Eliminación de Outliers
- Eliminación de columnas
- Creación de columnas nuevas
- Gestión de valores nulos
- Y un largo etcétera de técnicas que como Data Scientist hayas considerado las mejores para tu dataset.

### 2. Carga los datos de `diamonds_test.csv` para predecir.

**¿De dónde saco `diamonds_test.csv`?**

In [52]:
X_pred = pd.read_csv("diamonds_test.csv", index_col = 0)
X_pred.head()

FileNotFoundError: [Errno 2] No such file or directory: 'diamonds_test.csv'

In [18]:
X_pred.shape

(13449, 9)

**¿Por qué da error?** 

In [19]:
predictions_submit = model.predict(X_pred)
predictions_submit

ValueError: could not convert string to float: 'Ideal'

## IMPORTANTE: APLICAR LO MISMO A ESTOS DATOS QUE HAYÁIS APLICADO A LOS DATOS DE ENTRENAMIENTO

- SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 4 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS
- SI AL ARRAY CON EL QUE HICISTEIS `.fit()` LO NORMALIZASTEIS, PARA `.predict()` DEBÉIS NORMALIZARLO
- TODO IGUAL SALVO BORRAR FILAS, EL NÚMERO DE ROWS SE DEBE MANTENER EN ESTE SET, PUES LA PREDICCIÓN DEBE TENER 13449 FILAS, SI O SI

**Entonces, si al cargar los datos de train usé `index_col=0` para que utilizara la primera columna del conjunto de datos como índice, ¿tendré que hacerlo también para el conjunto `diamonds_test.csv`?**

In [ ]:
# ¿Qué opináis?
# Sí

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

In [20]:
X_pred = pd.read_csv("diamonds_test.csv", index_col=0)
X_pred.head()

carat        cut color clarity  depth  table     x     y     z
0   0.30      Ideal     H     SI2   60.0   56.0  4.41  4.43  2.65
1   0.34      Ideal     D      IF   62.1   57.0  4.52  4.46  2.79
2   1.57  Very Good     I     VS2   60.3   58.0  7.58  7.55  4.56
3   0.31      Ideal     H     VS2   61.8   57.0  4.32  4.36  2.68
4   1.51       Good     I    VVS1   64.0   60.0  7.26  7.21  4.63

**¿Qué más habrá que quitar o hacer?**

In [ ]:
# ¿Qué opináis?
# En este caso, quitar todas las columnas, menos la x, y y z

![](https://www.tierraljelechu.com/web/img/wiki_up/1.996-SorpresaDto.-1-Red.jpg)

**Truqui**: Carga X que debe estar actualmente actualizada a cómo la usaste para entrenar a `model` y haz que `X_pred` sea igual. ¡Sin quitar filas!

In [21]:
X

array([[6.73, 6.7 , 4.23],
       [4.14, 4.17, 2.66],
       [4.86, 4.82, 2.96],
       ...,
       [5.33, 5.34, 3.32],
       [6.47, 6.56, 3.88],
       [5.33, 5.27, 3.2 ]])

In [22]:
X_pred = np.array(X_pred[['x','y','z']])

Porque:
    - SI EL ARRAY CON EL QUE HICISTEIS `.fit()` ERA DE 3 COLUMNAS, PARA `.predict()` DEBEN SER LAS MISMAS

In [23]:
X_pred

array([[4.41, 4.43, 2.65],
       [4.52, 4.46, 2.79],
       [7.58, 7.55, 4.56],
       ...,
       [7.3 , 7.37, 4.53],
       [7.79, 7.68, 4.66],
       [4.29, 4.34, 2.64]])

### 3. AHORA puedo hacer la predicción que será lo que subirás a Kaggle. 

**¿Qué es lo que subirás a Kaggle?**

In [ ]:
# respuestas

In [24]:
predictions_submit = model.predict(X_pred)
predictions_submit

array([ -260.20029487,   116.70396517,  9699.26794807, ...,
        8903.59284445, 10336.93013348,  -601.00685048])

**¡PERO! Para subir a Kaggle la predicción, ésta tendrá que tener una forma específica y no valdrá otra.**

En este caso, la **MISMA** forma que `sample_submission.csv`. 

**¿De dónde saco `sample_submission.csv`?**

In [25]:
sample = pd.read_csv("sample_submission.csv")

In [26]:
sample.head()

id  price
0   0  12132
1   1  11786
2   2  14684
3   3  15425
4   4   6724

In [27]:
sample.shape

(13449, 2)

### 4. Mete tus predicciones en un dataframe. 

En este caso, la **MISMA** forma que `sample_submission.csv`. 

In [3]:
submission = pd.DataFrame({"id": range(len(predictions_submit)), "price": predictions_submit})

NameError: name 'pd' is not defined

In [2]:
submission.head()

NameError: name 'submission' is not defined

In [30]:
submission.shape

(13449, 2)

### 5. Pásale el CHEQUEATOR para comprobar que efectivamente está listo para subir a Kaggle.

In [31]:
def chequeator(df_to_submit):
    """
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.
    """
    if df_to_submit.shape == sample.shape:
        if df_to_submit.columns.all() == sample.columns.all():
            if df_to_submit.id.all() == sample.id.all():
                print("You're ready to submit!")
                submission.to_csv("submission.csv", index = False) #muy importante el index = False
                urllib.request.urlretrieve("https://i.kym-cdn.com/photos/images/facebook/000/747/556/27a.jpg", "gfg.png")     
                img = Image.open("gfg.png")
                img.show()   
            else:
                print("Check the ids and try again")
        else:
            print("Check the names of the columns and try again")
    else:
        print("Check the number of rows and/or columns and try again")
        print("\nMensaje secreto de Clara: No me puedo creer que después de todo este notebook hayas hecho algún cambio en las filas de `diamonds_test.csv`. Lloro.")


In [32]:
help(chequeator)

Help on function chequeator in module __main__:

chequeator(df_to_submit)
    Esta función se asegura de que tu submission tenga la forma requerida por Kaggle.
    
    Si es así, se guardará el dataframe en un `csv` y estará listo para subir a Kaggle.
    
    Si no, LEE EL MENSAJE Y HAZLE CASO.
    
    Si aún no:
    - apaga tu ordenador, 
    - date una vuelta, 
    - enciendelo otra vez, 
    - abre este notebook y 
    - leelo todo de nuevo. 
    Todos nos merecemos una segunda oportunidad. También tú.



In [34]:
chequeator(submission)

You're ready to submit!
